### DISCUSSION TITLE
'WorkflowDefinitionVersion' can be separated?

### DISCUSSION THREAD

**Discussion by jxnkwlp** [elsa-workflows/elsa-core#709] [Ideas]
In persistence , 'WorkflowDefinitionVersion' contain workflow definition and all versions , can it split by definition and versions ?

consider reason:  if there has many items in 'WorkflowDefinitionVersion' table , it may be list slow when frontend show .

**Comment by sfmskywalker** [comment]
It's possible, and we may still consider doing so if it turns out to be a real problem for Elsa 2. We'll keep an eye on that.

**Comment by DavidPx** [comment]
If you're using one of the EntityFramework providers adding an index to the definition table would help quite a bit - there are no indexes on any of those tables right now!

@sfmskywalker , is this something you'd like to see?  [Here is a proof of concept](https://github.com/DavidPx/elsa-core/commit/bb31d157102a62818ff7f3448beea10d9bf7a4ba).  

My index goes: Status, DefinitionId, Version.  In my use case I plan on showing my users their active/completed work in an area that will contain several different workflows.  I included Version so that I can exclude old, outdated tasks if needed.  I put Status in front because it should help the performance of the `ContinueRunningWorkflows` startup task.

**Reply by sfmskywalker** [reply]
Yep, we definitely need to add indexes for the fields on which we query 👍

**Reply by DavidPx** [reply]
I'm about to add those indicies to the EF Core configuration classes but could use some guidance on the Boookmark entity - I don't understand the use cases on that table.  A summary of what I plan on doing - please correct as is needed!

Workflow Instance:

* Status, DefinitionId, Version

Workflow Definition:

* DefinitionId, Version

Bookmark:

* ActivityType, TenantId, Hash

**Comment by sfmskywalker** [comment]
Looking good. I would go a bit further even and add indices for the following:

**Workflow Definition**

* TenantId
* Version
* Name
* IsLatest
* IsPublished

**Workflow Instance**

* TenantId
* DefinitionId
* WorkflowStatus
* CorrelationId
* ContextType
* ContextId
* Name
* CreatedAt
* LastExecutedAt
* FinishedAt
* CancelledAt
* FaultedAt

**Bookmark**

* TenantId
* Hash
* ActivityType
* ActivityId
* WorkflowInstanceId

Most of the above fields are being queried or ordered by, and some of them I suspect will eventually be used in a query.

**Reply by DavidPx** [reply]
To clarify - are you saying that there should be one index for each column you've listed here?  

For example, in workflow definition we'd have:
IX_WorkflowDefinition_TenantId
IX_WorkflowDefinition_Version
IX_WorkflowDefinition_Name
IX_WorkflowDefinition_IsLatest
IX_WorkflowDefinition_IsPublished

**Reply by sfmskywalker** [reply]
Yes exactly :+1:

**Reply by DavidPx** [reply]
I hit a snag while creating the migration for SQL Server.  SQL Server doesn't allow indicies on `nvarchar(max)` fields so the EF changed the datatypes of the affected string fields to `nvarchar(450)`.  While I'm using SQL Server 2019, that appears to be a compatibility selection to pre-2016 versions which capped the index size to 900 bytes.

Here is the relevant migration:  https://github.com/DavidPx/elsa-core/commit/a5279907bf52327150ab7566dd8817d62638d9b0

What are your thoughts on this?

**Reply by sfmskywalker** [reply]
Which field does this affect? I’m pretty sure we don’t need nvarchar(max) for any of the fields to be indexed, so would be fine to update the entity configurations as necessary.

**Reply by DavidPx** [reply]
See the commit I linked - essentially all of the string fields that you listed in your earlier response.

I would agree that 450 characters is plenty for things that we're indexing.

**Reply by sfmskywalker** [reply]
Ah I see. Yeah I don’t see an issue with this. Of course strictly speaking of the persistence provider imposes a max length on fields, one might argue that we should enforce some limit in the domain itself, but doesn’t sound worth the trouble to me.
